File: scripts/run-pipeline.ts
#!/usr/bin/env node
/**
 * 本地运行 Pipeline 脚本
 * 
 * 简化本地运行 Pipeline 的流程，支持通过命令行参数传递输入
 * 
 * 用法:
 *   npm run pipeline -- BuildPipeline
 *   npm run pipeline -- FlutterBuildPipeline --build-command "flutter build"
 *   npm run pipeline -- BuildPipeline --setup-command "npm install" --build-command "npm run build"
 * 
 * 选项:
 *   <PipelineName>           Pipeline 类名（必需）
 *   --<input-key> <value>    设置输入参数（如 --build-command "npm run build"）
 *   --help                   显示帮助信息
 */

import * as path from 'path';
import * as fs from 'fs';

// 解析命令行参数
const args = process.argv.slice(2);

if (args.length === 0 || args[0] === '--help' || args[0] === '-h') {
  console.log(`
本地运行 Pipeline 脚本

用法:
  npm run pipeline -- <PipelineName> [选项]

示例:
  # 运行 BuildPipeline（使用默认参数）
  npm run pipeline -- BuildPipeline

  # 运行 BuildPipeline 并设置构建命令
  npm run pipeline -- BuildPipeline --build-command "npm run build"

  # 运行 FlutterBuildPipeline
  npm run pipeline -- FlutterBuildPipeline

  # 设置多个参数
  npm run pipeline -- BuildPipeline \\
    --setup-command "npm install" \\
    --build-command "npm run build" \\
    --artifact-path "dist/**"

选项:
  --<input-key> <value>    设置输入参数
  --help, -h               显示此帮助信息

可用的 Pipeline:
  - BuildPipeline
  - FlutterBuildPipeline
  - ReleasePipeline
  - VersionBumpPipeline
`);
  process.exit(0);
}

const pipelineName = args[0];
const inputArgs: Record<string, string> = {};

// 解析输入参数
for (let i = 1; i < args.length; i += 2) {
  const key = args[i];
  const value = args[i + 1];
  
  if (key && key.startsWith('--')) {
    const inputKey = key.substring(2).replace(/-/g, '-');
    if (value) {
      inputArgs[inputKey] = value;
    }
  }
}

// 设置环境变量（GitHub Actions 格式）
for (const [key, value] of Object.entries(inputArgs)) {
  const envKey = `INPUT_${key.toUpperCase().replace(/-/g, '_')}`;
  process.env[envKey] = value;
}

// 动态加载 Pipeline 类
async function loadPipeline(pipelineName: string): Promise<any> {
  // 尝试从编译后的文件加载
  const possiblePaths = [
    path.join(process.cwd(), 'dist', 'src', 'pipelines', 'base', `${pipelineName.toLowerCase().replace('pipeline', '')}-pipeline.js`),
    path.join(process.cwd(), 'dist', 'src', 'pipelines', 'build', `${pipelineName.toLowerCase().replace('pipeline', '')}-pipeline.js`),
    path.join(process.cwd(), 'dist', 'src', 'pipelines', 'test', `${pipelineName.toLowerCase().replace('pipeline', '')}-pipeline.js`),
  ];

  // 也尝试直接按类名查找
  const allPipelineFiles: string[] = [];
  
  function findPipelineFiles(dir: string): void {
    if (!fs.existsSync(dir)) return;
    
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        findPipelineFiles(fullPath);
      } else if (entry.isFile() && entry.name.endsWith('.js') && entry.name.includes('pipeline')) {
        allPipelineFiles.push(fullPath);
      }
    }
  }

  findPipelineFiles(path.join(process.cwd(), 'dist', 'src', 'pipelines'));

  // 尝试加载
  for (const filePath of [...possiblePaths, ...allPipelineFiles]) {
    if (fs.existsSync(filePath)) {
      try {
        const module = require(filePath);
        if (module[pipelineName]) {
          return module[pipelineName];
        }
      } catch (e) {
        // 继续尝试下一个
      }
    }
  }

  throw new Error(`无法找到 Pipeline 类: ${pipelineName}\n请确保已运行 npm run build`);
}

// 运行 Pipeline
async function main() {
  try {
    console.log(`🚀 运行 Pipeline: ${pipelineName}\n`);
    
    if (Object.keys(inputArgs).length > 0) {
      console.log('📋 输入参数:');
      for (const [key, value] of Object.entries(inputArgs)) {
        console.log(`   ${key}: ${value}`);
      }
      console.log('');
    }

    const PipelineClass = await loadPipeline(pipelineName);
    const pipeline = new PipelineClass(inputArgs);
    
    console.log('⏳ 执行中...\n');
    const result = await pipeline.run();
    
    console.log('\n📊 执行结果:');
    console.log(JSON.stringify(result, null, 2));
    
    if (result.success) {
      console.log('\n✅ Pipeline 执行成功！');
      process.exit(0);
    } else {
      console.log('\n❌ Pipeline 执行失败！');
      process.exit(result.exitCode || 1);
    }
  } catch (error: any) {
    console.error('\n❌ 错误:', error.message);
    if (error.stack) {
      console.error(error.stack);
    }
    process.exit(1);
  }
}

main();

File: python/scripts/run_pipeline.py
#!/usr/bin/env python3
"""
本地运行 Pipeline 脚本

简化本地运行 Pipeline 的流程，支持通过命令行参数传递输入

用法:
  python scripts/run_pipeline.py BuildPipeline
  python scripts/run_pipeline.py FlutterBuildPipeline --build-command "flutter build"
  python scripts/run_pipeline.py BuildPipeline --setup-command "npm install" --build-command "npm run build"

选项:
  <PipelineName>           Pipeline 类名（必需）
  --<input-key> <value>    设置输入参数（如 --build-command "npm run build"）
  --help                   显示帮助信息
"""

import sys
import os
import argparse
import importlib.util
from pathlib import Path


def load_pipeline(pipeline_name: str):
    """动态加载 Pipeline 类"""
    # 添加项目根目录到路径
    project_root = Path(__file__).parent.parent
    if str(project_root) not in sys.path:
        sys.path.insert(0, str(project_root))
    
    # 尝试从可能的路径加载
    possible_paths = [
        project_root / "src" / "pipelines" / "base" / f"{pipeline_name.lower().replace('pipeline', '')}_pipeline.py",
        project_root / "src" / "pipelines" / "build" / f"{pipeline_name.lower().replace('pipeline', '')}_pipeline.py",
        project_root / "src" / "pipelines" / "test" / f"{pipeline_name.lower().replace('pipeline', '')}_pipeline.py",
    ]
    
    # 也尝试直接按类名查找
    all_pipeline_files = []
    pipelines_dir = project_root / "src" / "pipelines"
    if pipelines_dir.exists():
        for file_path in pipelines_dir.rglob("*.py"):
            if not file_path.name.startswith("_") and "pipeline" in file_path.name.lower():
                all_pipeline_files.append(file_path)
    
    # 尝试加载
    for file_path in [*possible_paths, *all_pipeline_files]:
        if file_path.exists():
            try:
                # 计算模块路径（相对于项目根目录）
                relative_path = file_path.relative_to(project_root)
                module_path = str(relative_path.with_suffix("")).replace("/", ".").replace("\\", ".")
                
                spec = importlib.util.spec_from_file_location(module_path, file_path)
                if spec is None or spec.loader is None:
                    continue
                
                module = importlib.util.module_from_spec(spec)
                sys.modules[module_path] = module
                spec.loader.exec_module(module)
                
                if hasattr(module, pipeline_name):
                    return getattr(module, pipeline_name)
            except Exception as e:
                continue
    
    raise ValueError(f"无法找到 Pipeline 类: {pipeline_name}")


def main():
    parser = argparse.ArgumentParser(description="本地运行 Pipeline")
    parser.add_argument("pipeline", help="Pipeline 类名")
    
    # 解析已知参数，剩余的参数作为输入参数
    args, remaining = parser.parse_known_args()
    
    # 解析输入参数
    input_args = {}
    i = 0
    while i < len(remaining):
        arg = remaining[i]
        if arg.startswith("--") and i + 1 < len(remaining):
            key = arg[2:].replace("-", "-")
            value = remaining[i + 1]
            input_args[key] = value
            i += 2
        else:
            i += 1
    
    # 设置环境变量（GitHub Actions 格式）
    for key, value in input_args.items():
        env_key = f"INPUT_{key.upper().replace('-', '_')}"
        os.environ[env_key] = value
    
    try:
        print(f"🚀 运行 Pipeline: {args.pipeline}\n")
        
        if input_args:
            print("📋 输入参数:")
            for key, value in input_args.items():
                print(f"   {key}: {value}")
            print()
        
        # 添加项目根目录到路径
        script_dir = Path(__file__).parent.parent
        project_root = script_dir.parent  # python/ 的父目录是项目根目录
        
        if str(script_dir) not in sys.path:
            sys.path.insert(0, str(script_dir))
        
        PipelineClass = load_pipeline(args.pipeline)
        
        # 使用项目根目录创建 Pipeline
        pipeline = PipelineClass(input_args, project_root=str(project_root))
        
        print("⏳ 执行中...\n")
        result = pipeline.run()
        
        print("\n📊 执行结果:")
        print(f"   Success: {result.success}")
        print(f"   Message: {result.message}")
        print(f"   Exit Code: {result.exit_code}")
        if result.data:
            print(f"   Data: {result.data}")
        
        if result.success:
            print("\n✅ Pipeline 执行成功！")
            sys.exit(0)
        else:
            print("\n❌ Pipeline 执行失败！")
            sys.exit(result.exit_code or 1)
    except Exception as e:
        print(f"\n❌ 错误: {str(e)}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()

